flink-connectors/flink-connector-pubsub/src/main/java/org/apache/flink/streaming/connectors/pubsub/PubSubSource.java

```diff
@@ -0,0 +1,294 @@
+import com.google.cloud.pubsub.v1.AckReplyConsumer;
+import com.google.cloud.pubsub.v1.Subscriber;
+import com.google.pubsub.v1.ProjectSubscriptionName;
+import com.google.pubsub.v1.PubsubMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.cloud.pubsub.v1.SubscriptionAdminSettings.defaultCredentialsProviderBuilder;
+
+/**
+ * PubSub Source, this Source will consume PubSub messages from a subscription and Acknowledge them on the next checkpoint.
+ * This ensures every message will get acknowledged at least once.
+ */
+public class PubSubSource<OUT> extends MultipleIdsMessageAcknowledgingSourceBase<OUT, String, AckReplyConsumer>
+		implements ResultTypeQueryable<OUT>, ParallelSourceFunction<OUT>, StoppableFunction {
+	private static final Logger LOG = LoggerFactory.getLogger(PubSubSource.class);
+	protected DeserializationSchema<OUT> deserializationSchema;
+	protected SubscriberWrapper subscriberWrapper;
+
+	protected boolean running = true;
+	protected transient volatile SourceContext<OUT> sourceContext = null;
+
+	protected PubSubSource() {
+		super(String.class);
+	}
+
+	protected void setDeserializationSchema(DeserializationSchema<OUT> deserializationSchema) {
+		this.deserializationSchema = deserializationSchema;
+	}
+
+	protected void setSubscriberWrapper(SubscriberWrapper subscriberWrapper) {
+		this.subscriberWrapper = subscriberWrapper;
+	}
+
+	@Override
+	public void open(Configuration configuration) throws Exception {
+		super.open(configuration);
+		subscriberWrapper.initialize();
+		if (hasNoCheckpointingEnabled(getRuntimeContext())) {
+			throw new IllegalArgumentException("The PubSubSource REQUIRES Checkpointing to be enabled and " +
+				"the checkpointing frequency must be MUCH lower than the PubSub timeout for it to retry a message.");
+		}
+
+		getRuntimeContext().getMetricGroup().gauge("PubSubMessagesProcessedNotAcked", this::getOutstandingMessagesToAck);
+		getRuntimeContext().getMetricGroup().gauge("PubSubMessagesReceivedNotProcessed", subscriberWrapper::amountOfMessagesInBuffer);
+	}
+
+	private boolean hasNoCheckpointingEnabled(RuntimeContext runtimeContext) {
+		return !(runtimeContext instanceof StreamingRuntimeContext && ((StreamingRuntimeContext) runtimeContext).isCheckpointingEnabled());
+	}
+
+	@Override
+	protected void acknowledgeSessionIDs(List<AckReplyConsumer> ackReplyConsumers) {
+		ackReplyConsumers.forEach(AckReplyConsumer::ack);
+	}
+
+	@Override
+	public void run(SourceContext<OUT> sourceContext) throws Exception {
+		this.sourceContext = sourceContext;
+		subscriberWrapper.start();
+
+		while (subscriberWrapper.isRunning()) {
+			try {
+				Optional.ofNullable(subscriberWrapper.take())
+						.ifPresent(this::processMessage);
+			} catch (InterruptedException e) {
+				LOG.debug("Interrupted - stop or cancel called?");
+			}
+		}
+
+		nackOutstandingMessages();
```

Comment:
The pubsub connector doesn't like to get shutdown with unacknowledged messages.

So before I added this, whenever I would cancel a flink job there would be unacknowledged messages. And while the flink job would shutdown the job and start unloading classes from the classloader the PubSub sdk would still be doing stuff in the background and eventually start throwing ClassNotFound exceptions.

It also seems cleaner to nack messages when shutting down gracefully. In any way, if the nack's don't happen, PubSub will resend the messages eventually so there is no risk of not guaranteeing atleast once.

Reply:
But isn't the shutdown being delayed (and thus recovery times go up) if we need to ack a lot of outstanding messages.
Isn't there a better solution to ensure that PubSub has been shut down before Flink starts unloading classes?
(If there are PubSub threads being leaked, that's something we should investigate and understand.)

Reply:
Yes this would slow down the shutdown, in one of our cases saw ~20k messages that has to be nacked and the slow down was in the order of 100s of milliseconds to a second, which sounds acceptable?

Regarding if this is the best solution: I've asked the sdk owners this question here: https://github.com/googleapis/google-cloud-java/issues/3752 and they actually implemented this behavior not too long ago and don't intend on changing it. 

Threads are only leaked if we don't wait for the subscriber to shutdown.`subscriberWrapper.awaitTerminated();` makes sure the subscriber is completely done and terminated.

Reply:
Okay. Then let's keep this as is.

Reply:
The nackOutstandingMessages() is gone now (because of sync api) I still have the `awaitTerminated()` call to prevent leaking threads

Reply:
Small note, I am struggling to get this shutting down working, it seems to either leak threads or is still shutting down threads eventhough `awaitTerminated()` and `isTerminated()` both say it should be all done.

To reproduce, simple stop or cancel the job and see the output of a taskmanager, it will throw class not found exceptions.

Reply:
I've created this issue over at the pubsub sdk github: https://github.com/googleapis/google-cloud-java/issues/4558

I do wonder if this isn't a bug in flink as well. According to the docs: https://ci.apache.org/projects/flink/flink-docs-release-1.7/monitoring/debugging_classloading.html#unloading-of-dynamically-loaded-classes classes should not be unloaded if there are still non gc'd objects that use it.

Reply:
Afaik, the JVM should not unload classes if they are still referenced by something (in this case a running thread). 
So even if Flink has lost it's reference to the Classloader, the classes should still be present till the thread has ceased to exist.
Which classes are not found in these shutdown scenarios?

Reply:
Yeah, that's what I was thinking after reading the Flink docs. The exceptions I am seeing always seems to be constructors or similar of different classes.

I think the following is happening: the background thread creates new objects after the job has finished (and thus class unloading has begun) These new objects might not have been referenced at some point and thus have been unloaded. 

In the google issue there's an code sample to wait for the background threads to shutdown, I'll give that a try.

Reply:
Alright, I've got the shutdown working cleanly now (just pushed).

The last point I would like your advice on is this line: https://github.com/apache/flink/pull/6594/files#diff-5d1317f139d297f4e67ff13defa4d66cR208

I swallow the InterruptionException here because I see the following behavior:

When the job gets stopped (StoppableFunction) stop() gets called and everything shutsdown gracefully. No problem here.

When the job gets cancelled, cancel() gets called and immediatly afterwards the thread that executed run() gets an interrupt. If after 30 seconds the thread still has not shutdown it will interrupt again and again after 3 times it will kill the taskmanager jvm.

Because the first interrupt happens immediatly and we still want the threads to shutdown cleanly I opt to simply swallow the exception here as we know the threads are shutting down at this point. Should we keep it as simple as this or add some logic to only ignore the first interrupt?